### PR TITLE
[FIX] web: fix race condition with m2o quick create

### PR DESCRIPTION
**Pull request by aab-odoo**
Let's assume a form view containing a many2one with an onchange
that updates the value of a one2many. Do a quick create in the
many2one. While the name_create request is pending, add a row to
the one2many but do not leave it. When the name_create returns, and
the onchange has been performed, the one2many is reset, and the row
is no longer in edition (worse, it could be invalid, i.e. in a state
that the user could not have reached in a normal situation).

This commit fixes this issue by considering the whole [name_create +
onchange] operation as one. This operation is executed in the mutex
of the model, so the other request (adding a row to the one2many) is
delayed until the many2one value has been correctly set.

This fixes an issue with the sale and rental tours (on sale_order),
that had been deactivated for a while.

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 1]
```diff
@@ -1523,18 +1523,40 @@ var BasicModel = AbstractModel.extend({
...
+        if (!data || (!data.id && !data.display_name)) {
             record._changes[fieldName] = false;
             return Promise.resolve();
         }
 
+        const field = record.fields[fieldName];
+        const coModel = field.type === 'reference' ? data.model : field.relation;
+        if (field.type === 'many2one' && !data.id && data.display_name) {
+            // only display_name given -> do a name_create
+            await this._rpc({
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: stylistic change:
```js
const result = await this._rpc({...});
data = {id: result[0], display_name: result[1];
```

File: addons/web/static/tests/fields/relational_fields/field_one2many_tests.js [hunk 2]
```diff
@@ -8741,6 +8741,56 @@ QUnit.module('fields', {}, function () {
             form.destroy();
         });
 
+        QUnit.test('one2many reset by onchange (of another field) while being edited', async function (assert) {
+            // In this test, we have a many2one and a one2many. The many2one has an onchange that
+            // updates the value of the one2many. We set a new value to the many2one (name_create)
+            // such that the onchange is delayed. During the name_create, we click to add a new row
```
Comment (ged-odoo) [inline addons/web/static/tests/fields/relational_fields/field_one2many_tests.js]: nice to have a description here

File: addons/sale_product_configurator/static/src/js/product_configurator_widget.js [hunk 3]
```diff
@@ -33,7 +33,7 @@ ProductConfiguratorWidget.include({
      * @override
      * @private
      */
-    _onFieldChanged: function (ev) {
+    reset: function (record, ev) {
         this.restoreProductTemplateId = this.recordData.product_template_id;
```
Comment (aab-odoo) [inline addons/sale_product_configurator/static/src/js/product_configurator_widget.js]: if evtarget this

**Comment by ged-odoo** [comment]
robodoo r+ rebase-ff

**Comment by d-fence** [comment]
:1st_place_medal: Great

**Pull request closed without merging**